### PR TITLE
csr approver: choose between logging and status updates on errors

### PR DESCRIPTION
**Pull request by stlaz**
Add a noop operator client for generic controllers that are a part
of an operand rather than an operator.

### CODE REVIEW

File: pkg/operator/v1helpers/helpers.go [hunk 1]
```diff
@@ -21,6 +22,41 @@ import (
 	operatorv1 "github.com/openshift/api/operator/v1"
 )
 
+var _ OperatorClient = &NoOpOperatorClient{}
```
Comment (s-urbaniak) [inline pkg/operator/v1helpers/helpers.go:25]: :+1:

File: pkg/operator/v1helpers/helpers.go [hunk 2]
```diff
@@ -21,6 +22,41 @@ import (
...
+func (c *NoOpOperatorClient) GetObjectMeta() (*metav1.ObjectMeta, error) {
+	return &metav1.ObjectMeta{}, nil
+}
+
+func (c *NoOpOperatorClient) Informer() cache.SharedIndexInformer {
+	return nil
+}
+
+func (c *NoOpOperatorClient) GetOperatorState() (*operatorv1.OperatorSpec, *operatorv1.OperatorStatus, string, error) {
+	return &operatorv1.OperatorSpec{}, &operatorv1.OperatorStatus{}, "", nil
```
Comment (s-urbaniak) [inline pkg/operator/v1helpers/helpers.go:49]: is there any reason we return empty structs vs nil?

Comment (stlaz) [reply pkg/operator/v1helpers/helpers.go:49]: it's safer as the error is `nil` and so the controllers might assume (rightfully) that the returned structures exist

**Review by slaskawi — APPROVED** [review]

**Comment by stlaz** [comment]
/assign @deads2k 
for approval

**Comment by sttts** [comment]
It's not clear to me why we need this.

File: pkg/operator/v1helpers/helpers.go [hunk 3]
```diff
@@ -21,6 +22,41 @@ import (
...
+}
+
+func (c *NoOpOperatorClient) UpdateOperatorSpec(_ string, _ *operatorv1.OperatorSpec) (*operatorv1.OperatorSpec, string, error) {
+	return &operatorv1.OperatorSpec{}, "", nil
+}
+
+func (c *NoOpOperatorClient) UpdateOperatorStatus(_ string, _ *operatorv1.OperatorStatus) (*operatorv1.OperatorStatus, error) {
```
Comment (deads2k) [inline pkg/operator/v1helpers/helpers.go:56]: I'd like to see this method print the attempted state change at info level if conditions are "good" and at warning level with events if conditions are "bad".

The control loops that accept this as an interface expect to report status via this function today.

File: pkg/operator/v1helpers/helpers.go [hunk 4]
```diff
@@ -21,6 +22,41 @@ import (
...
+
+func (c *NoOpOperatorClient) Informer() cache.SharedIndexInformer {
+	return nil
+}
+
+func (c *NoOpOperatorClient) GetOperatorState() (*operatorv1.OperatorSpec, *operatorv1.OperatorStatus, string, error) {
+	return &operatorv1.OperatorSpec{}, &operatorv1.OperatorStatus{}, "", nil
+}
+
+func (c *NoOpOperatorClient) UpdateOperatorSpec(_ string, _ *operatorv1.OperatorSpec) (*operatorv1.OperatorSpec, string, error) {
```
Comment (deads2k) [inline pkg/operator/v1helpers/helpers.go:52]: if it only accepts status updates, it's ok, but this method should fail imo

**Comment by sttts** [comment]
/hold

I don't think a noop client without persistence implements the interface of an operator client.

**Review by slaskawi — APPROVED** [review]
/lgtm

File: pkg/operator/v1helpers/helpers.go [hunk 5]
```diff
@@ -14,13 +14,44 @@ import (
...
+}
+
+func (c *LoggingStatusUpdater) GetOperatorState() (*operatorv1.OperatorSpec, *operatorv1.OperatorStatus, string, error) {
+	return &operatorv1.OperatorSpec{}, &operatorv1.OperatorStatus{
+		Conditions: c.cachedConditions,
+	}, "", nil
+}
+
+func (c *LoggingStatusUpdater) UpdateOperatorStatus(_ string, status *operatorv1.OperatorStatus) (*operatorv1.OperatorStatus, error) {
+	klog.V(2).Info("received a request to update status, diff: %s", diff.ObjectDiff(status, &operatorv1.OperatorStatus{Conditions: c.cachedConditions}))
```
Comment (s-urbaniak) [inline pkg/operator/v1helpers/helpers.go:47]: .oO(i wish we could inject the logger at initialization time) ... actually ... does that make sense, potentially just a logging function?

File: pkg/operator/v1helpers/helpers.go [hunk 6]
```diff
@@ -14,13 +14,44 @@ import (
...
 	configv1 "github.com/openshift/api/config/v1"
 	operatorv1 "github.com/openshift/api/operator/v1"
 )
 
+var _ OperatorStatusUpdater = &LoggingStatusUpdater{}
+
+type LoggingStatusUpdater struct {
+	// cachedConditions store the last observed conditions to be returned in the
+	// GetOperatorState method in order to prevent excessive conditions update logging
+	cachedConditions []operatorv1.OperatorCondition
```
Comment (s-urbaniak) [inline pkg/operator/v1helpers/helpers.go:31]: we don't have to provide any synchronization for reads/writes? Here, `GetOperatorState` and `UpdateOperatorStatus` must not be used concurrently. We should at least make this clear in the constructor.

Comment (stlaz) [reply pkg/operator/v1helpers/helpers.go:31]: you're right, we should

File: pkg/operator/v1helpers/helpers.go (hunk 5 above)
Comment (stlaz) [reply pkg/operator/v1helpers/helpers.go:47]: `klog` is the logger all our operators use, I believe, and is heavily used in other controllers. I'd maybe do that in a follow-up if such a need arises.

**Pull request merged**